### DISCUSSION TITLE
Permit `foreach (ref T...)` loops on arrays

### DISCUSSION THREAD

**Discussion by MichalPetryka** [dotnet/csharplang#6962] [Language Ideas] [answered]
Permit by ref loops on arrays:
```C#
public static void A(int[] array)
{
    foreach (ref int r in array)
    {
        r = r + 1;
    }
}
```
This currently gives: `error CS1510: A ref or out value must be an assignable variable`.

Same code works on spans:
```C#
public static void S(Span<int> array)
{
    foreach (ref int r in array)
    {
        r = r + 1;
    }
}
```

[SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKBuIGYACY8pRgVwDtdsAzGZqUYAVGLgw0A3jUazmTFm2IpGAQQAUAS04YA2gF1G2KFGwBPAJQy506nPuNe0GNjAALRuti9G2jIyhfTiMTcys7B1lbSMjAgF4AxgBqRnIAbmtIgF9MxhyI2VyGZlZmFQBldXKAB2xOAB4/AD4Q00tc6JinWFcPLxgfP0TtVrDc+06YuXjElPTxuXz7fKygA==)

**Comment by CyrusNajmabadi** [comment]
I'm not opposed to this.  But it's very marginal.  You can just update your code to be `foreach (ref int r in array.AsSpan())`.  Given the trivial source workaround, i would not invest lang/compiler time here.

**Comment by jaredpar** [comment, marked as answer]
I think the biggest issue here is we don't, at least to my knowledge, have a spec for what we intended to allow here. The `ref` iteration variable for `foreach` was added in [this PR](https://github.com/dotnet/roslyn/pull/25002). That has no spec, issue or statement of when it should / shouldn't be allowed. All we have to go on is the tests. 

That's not surprising though. Back in 2018 we weren't good about ensuring we had solid specs for all of the behaviors that we added. Maybe @agocke remembers more or can point me to a spec I'm missing. 

For me the first step is documenting what our current behavior is and what the motivations are. My _personal intuition_ is that at a high level if the `Current` member is `ref` returning then we should support a `ref` iteration variable. That gets a bit more detailed though when considering `ref readonly` returns, and arrays which tend to be a special case here. If the spec'ing of that naturally falls out for arrays then I don't have a problem with it being added (but I agree with other comments about whether it's the best use of resources). 

But first step is definitely documenting the current behavior.

Links:
- Existing proposal: https://github.com/dotnet/csharplang/issues/3542
- Confirmed no spec: https://github.com/dotnet/csharplang/issues/3209

**Reply by svick** [reply]
Note that there is a work-in-progress `ref` `foreach` spec at https://github.com/dotnet/csharpstandard/pull/213, but it currently does not say anything about what types are supported in `ref` `foreach`.

**Reply by agocke** [reply]
FWIW I would just accept https://github.com/dotnet/csharplang/issues/3542 and implement it, I think it's just an oversight.

**Reply by jaredpar** [reply]
I would rather define what we do / don't allow before accepting an arbitrary extension. Lacking that we're left asking questions like "does this work when `Current` is `ref readonly`?".

**Reply by agocke** [reply]
Someone would have to confirm, but I suspect that this works for Spans because `Current` is `ref` returning.

I'm proposing an exception for arrays simply because arrays are exceptional. There's special code in the spec just for arrays, so spec'ing arrays separately makes sense to me.
